cajita/src/Cajita_Halo.hpp

```diff
@@ -124,52 +249,41 @@ struct Max
 // which neighbor that value will come.
 struct Replace
 {
-    template <class ViewType, class BufferType>
-    KOKKOS_INLINE_FUNCTION void
-    operator()( const ViewType &view, const BufferType &buffer, const int i,
-                const int j, const int k, const int l ) const
-    {
-        view( i, j, k, l ) = buffer( i, j, k, l );
-    }
 };
 
 } // end namespace ScatterReduce
 
 //---------------------------------------------------------------------------//
-// Halo exchange communication plan for migrating shared data between blocks.
+// General multiple array halo communication plan for migrating shared data
+// between blocks. Arrays may be defined on different entity types and have
+// different data types.
 //---------------------------------------------------------------------------//
-template <class Scalar, class DeviceType>
-class Halo
+template <class MemorySpace>
+class MultiHalo
 {
   public:
-    // Scalar type.
-    using value_type = Scalar;
-
-    // Device type.
-    using device_type = DeviceType;
-
-    // Execution space.
-    using execution_space = typename device_type::execution_space;
-
     // Memory space.
-    using memory_space = typename device_type::memory_space;
-
-    // View type.
-    using view_type = Kokkos::View<value_type ****, device_type>;
+    using memory_space = MemorySpace;
 
     /*!
       \brief Constructor.
-      \param layout The array layout to build the halo for.
-      \param pattern The halo pattern to use for halo communication
-      \param width An option halo cell width. The default is to use the entire
-      halo width of the block.
+      \tparam The arrays types to construct the halo for.
+      \param pattern The halo pattern to use for halo communication.
+      \param width Halo cell width. Must be less than or equal to the halo
+      width of the block.
+      \param arrays The arrays to build the halo for. These arrays must be
+      provided in the same order
     */
-    template <class ArrayLayout_t>
-    Halo( const ArrayLayout_t &layout, const HaloPattern &pattern,
-          const int width = -1 )
+    template <class... ArrayTypes>
+    MultiHalo( const HaloPattern &pattern, const int width,
+               ArrayTypes... arrays )
     {
-        // Duplicate the communicator so we have our own communication space.
-        MPI_Comm_dup( layout.localGrid()->globalGrid().comm(), &_comm );
+        // Get the MPI communicator. All arrays should have the same
+        // communicator.
```

Comment:
should -> must and document it

Reply:
I have documented this more and added an explicit check when I get the communicators by checking each array comm with `MPI_Comm_compare`

Reply:
I also added a check that the arrays have the same halo width.